FontCascade::width accepts a StringView

vitorroriz opened the pull request:
#### bb9b7e0a2862babdf94ad364dcbc3ce61c67f813
<pre>
FontCascade::width accepts a StringView
<a href="https://rdar.apple.com/151404312">rdar://151404312</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=293060">https://bugs.webkit.org/show_bug.cgi?id=293060</a>

Reviewed by Brent Fulgham.

There are a few cases where we instantiate a TextRun from a String because we want
to measure FontCascade::width and this function works with runs.

We are providing a wrapper that instantiates the TextRun from a StringView so
we don&apos;t have to do that on every call.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/inspector/InspectorOverlayLabel.cpp:
(WebCore::InspectorOverlayLabel::expectedSize):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayLineBuilder.cpp:
(WebCore::Layout::trailingEllipsisVisualRectAfterTruncation):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::hyphenWidth):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::width const):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebKit/WebProcess/WebCoreSupport/win/WebPopupMenuWin.cpp:
(WebKit::WebPopupMenu::setUpPlatformData):

Canonical link: <a href="https://commits.webkit.org/294986@main">https://commits.webkit.org/294986@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/248fcccbfee7a46e53cf4f5b5d0fbcb9bdc68701

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13709 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108877 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54339 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105727 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23733 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31935 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78791 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18410 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59125 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18219 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11583 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53715 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88002 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11642 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111266 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30843 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22742 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87781 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31204 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89736 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87431 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22266 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32312 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10037 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25200 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30771 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36074 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30565 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33900 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32126 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -151,14 +151,15 @@ class FontCascade final : public CanMakeWeakPtr<FontCascade>, public CanMakeChec
 
     float widthOfTextRange(const TextRun&, unsigned from, unsigned to, SingleThreadWeakHashSet<const Font>* fallbackFonts = nullptr, float* outWidthBeforeRange = nullptr, float* outWidthAfterRange = nullptr) const;
     WEBCORE_EXPORT float width(const TextRun&, SingleThreadWeakHashSet<const Font>* fallbackFonts = nullptr, GlyphOverflow* = nullptr) const;
+    WEBCORE_EXPORT float width(StringView, SingleThreadWeakHashSet<const Font>* fallbackFonts = nullptr, GlyphOverflow* = nullptr) const;
```
Since this method is never used with non-null `fallbackFonts` or `GlyphOverflow`, I wonder if we need to have our new method take these arguments?

vitorroriz replied:
That's a good point. I'll get rid of it until we need it.

vitorroriz replied:
Unless...do you think having it with the the extra arguments available force people to think if they need it or not?

brentfulgham reviewed (approved):
LGTM.

cdumez commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -303,6 +303,12 @@ float FontCascade::widthOfTextRange(const TextRun& run, unsigned from, unsigned
+float FontCascade::width(StringView text, SingleThreadWeakHashSet<const Font>* fallbackFonts, GlyphOverflow* glyphOverflow) const
+{
+    TextRun run { text };
+    return width ( CheckedRef { run }, fallbackFonts, glyphOverflow);
```
@rniwa `run` is a stack variable here, there is no way it can get destroyed while running `width()` so the CheckedRef doesn't seem useful here. Either we have a static analysis false positive or the smart pointer adoption here is speculative. If a false positive, we may consider silencing the warning instead.

@vitorroriz : There is an extra space between `width` and `(`.

vitorroriz replied:
That makes sense, I can check what the static analyzer says when passing a l-value reference without the CheckedRef. I was looking at [1] when doing this. Let me try a new iteration.

[1] https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/294967%40main%20%283475%29/scan-build-output/StaticAnalyzer/WebCore/StaticAnalyzerReports/report-FontCascade.h-widthOfSpaceString-1-d3dbd8.html#EndPath

vitorroriz committed:
FontCascade::width accepts a StringView
rdar://151404312
https://bugs.webkit.org/show_bug.cgi?id=293060

Reviewed by Brent Fulgham.

There are a few cases where we instantiate a TextRun from a String because we want
to measure FontCascade::width and this function works with runs.

We are providing a wrapper that instantiates the TextRun from a StringView so
we don't have to do that on every call.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/inspector/InspectorOverlayLabel.cpp:
(WebCore::InspectorOverlayLabel::expectedSize):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayLineBuilder.cpp:
(WebCore::Layout::trailingEllipsisVisualRectAfterTruncation):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::hyphenWidth):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::width const):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebKit/WebProcess/WebCoreSupport/win/WebPopupMenuWin.cpp:
(WebKit::WebPopupMenu::setUpPlatformData):

Canonical link: https://commits.webkit.org/294986@main
